### PR TITLE
Re-implement no-slip wall using node mask

### PR DESCRIPTION
**Pull request by neilmatula**
**Pull-request type:**
- [ ] Bug fix 
- [ ] Documentation update
- [ ] Feature enhancement

## Description of the pull-request

*Provide a description of your proposed changes. If there is a related issue, please specify.**

## Checklist

**NOTE** The checklist below is a suggestion and not mandatory. You don't have
to _check all boxes_ to submit a pull request. However, please add a brief
explanation in your pull request summary explaining the omission for the benefit
of reviewers and developers.

*All pull requests*
- [ ] Builds successfully (*must test on at least one system/compiler combination*)
  - Operating systems 
    - [ ] Linux
    - [ ] MacOS
  - Compilers 
    - [ ] GCC
    - [ ] LLVM/Clang
    - [ ] Intel compilers
    - [ ] NVIDIA CUDA
- [ ] Compiles without warnings
- [ ] Passes all unit tests
- [ ] Passes all regression tests
- [ ] Documentation builds without errors

*For new code updates*
- [ ] Documentation updates - additions to user, theory, and verification manuals
- [ ] New unit tests providing coverage for new code, bug fixes etc.
- [ ] New regression tests exercising the new code

### CODE REVIEW

File: reg_tests/test_files/ablNeutralEdgeNoSlip/README [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  - linear increasing at a rate of 3 K per 1000 m above 750 m
+
+Initial velocity:
+  - Constant 8 m/s from 245 degrees throughout with sinusoidal
+    perturbations near the surface
+
+Surface:
+  - Rough wall shear stress model based on Monin-Obukhov similarity.
+  - z0 = 0.2 m
+  - qWall = 0 W/m^2
```
Comment (sayerhs) [inline reg_tests/test_files/ablNeutralEdgeNoSlip/README:43]: Should we add an explanation of the _no-slip_ implementation here to differentiate this from the `ablNeutralEdge` case?

File: src/LowMachEquationSystem.C [hunk 2]
```diff
@@ -1177,6 +1179,12 @@ MomentumEquationSystem::register_nodal_fields(
     stk::mesh::put_field_on_mesh(*actuatorSourceLHS, *part, nDim, nullptr);
     stk::mesh::put_field_on_mesh(*g, *part, nullptr);
   }
+
+  GenericFieldType& node_mask =
+    realm_.meta_data().declare_field<GenericFieldType>(
+      stk::topology::NODE_RANK, "abl_wall_no_slip_wall_func_node_mask");
+  double one = 1;
+  stk::mesh::put_field_on_mesh(node_mask, *part, 1, &one);
```
Comment (sayerhs) [inline src/LowMachEquationSystem.C:1187]: Minor nitpick. This can just be `ScalarFieldType` as it is implemented on nodes and not on edges/elements.

File: src/edge_kernels/MomentumABLWallShearStressEdgeKernel.C [hunk 3]
```diff
@@ -25,27 +25,37 @@ namespace nalu {
...
+    meFC_(MasterElementRepo::get_surface_master_element<typename BcAlgTraits::FaceTraits>()),
+    meSCS_(MasterElementRepo::get_surface_master_element<typename BcAlgTraits::ElemTraits>())
 {
   faceDataPreReqs.add_cvfem_face_me(meFC_);
+  elemData.add_cvfem_surface_me(meSCS_);
 
   faceDataPreReqs.add_face_field(exposedAreaVec_, BcAlgTraits::numFaceIp_, BcAlgTraits::nDim_);
   faceDataPreReqs.add_face_field(wallShearStress_, BcAlgTraits::numFaceIp_, BcAlgTraits::nDim_);
+
+  elemData.add_coordinates_field(coordinates_, BcAlgTraits::nDim_, CURRENT_COORDINATES);
```
Comment (sayerhs) [inline src/edge_kernels/MomentumABLWallShearStressEdgeKernel.C:47]: It looks like coordinates is not used in the algorithm. Can eliminate this registration step, as well as tracking coordinates for this kernel.

File: src/edge_kernels/MomentumABLWallShearStressEdgeKernel.C [hunk 4]
```diff
@@ -54,7 +64,17 @@ MomentumABLWallShearStressEdgeKernel<BcAlgTraits>::execute(
...
+    // Decide which node the shear stress will be applied to
+    // rowBase is the first row in the matrix associated with the velocity for this node
+    int rowBase;
+    if (slip_) {
+      // For slip, we want the node on the wall
+      rowBase = ipNodeMap[ip] * BcAlgTraits::nDim_;
+    } else {
+      // For no-slip, we want the node opposite the node at the wall
+      rowBase = meSCS_->opposingNodes(elemFaceOrdinal, ip) * BcAlgTraits::nDim_;
+    }
```
Comment (sayerhs) [inline src/edge_kernels/MomentumABLWallShearStressEdgeKernel.C:77]:
```suggestion
    const int rowBase = slip_ 
      ? (ipNodeMap[ip] * BcAlgTraits::nDim_) 
      : (meSCS_->opposingNodes(elemFaceOrdinal, ip) * BcAlgTraits::nDim_);
```

File: src/ngp_algorithms/MomentumABLWallFuncMaskUtil.C [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+MomentumABLWallFuncMaskUtil::execute()
+{
+  using Traits = nalu_ngp::NGPMeshTraits<stk::mesh::NgpMesh>;
+
+  const auto& meta     = realm_.meta_data();
+  const auto ngpMesh   = realm_.ngp_mesh();
+  const auto& fieldMgr = realm_.ngp_field_manager();
+  const auto myNodeMask    = fieldMgr.get_field<double>(maskNodeIndex_);
+
+  // This selector only includes edges that touch the abl wall (including those edges on the wall)
```
Comment (sayerhs) [inline src/ngp_algorithms/MomentumABLWallFuncMaskUtil.C:38]: Outdated comment?

File: src/LowMachEquationSystem.C [hunk 6]
```diff
@@ -2051,20 +2059,32 @@ MomentumEquationSystem::register_wall_bc(
...
-            *this, *part, solverAlgMap, "wall_fcn");
+        std::tie(faceElemSolverAlg, solverAlgWasBuilt) =
+          build_or_add_part_to_face_elem_solver_alg(
+            wfAlgType, *this, *part, elemTopo, solverAlgMap, "wall_func");
 
-        ElemDataRequests& dataPreReqs = solverAlg->dataNeededByKernels_;
-        auto& activeKernels = solverAlg->activeKernels_;
+        if (userData.isNoSlip_) {
+          notProjectedPart_.push_back(part);
+          if( ablWallNodeMask_ == nullptr ) {
```
Comment (sayerhs) [inline src/LowMachEquationSystem.C:2072]:
```suggestion
          if (!ablWallNodeMask_) {
```

File: reg_tests/test_files/ablNeutralEdgeNoSlip/README (hunk 1 above)
Comment (psakievich) [reply reg_tests/test_files/ablNeutralEdgeNoSlip/README:37]: Yeah good idea. Maybe after the first sentence add something like:

_This case utilizes the no-slip wall stress boundary condition proposed by Jaegle et al. in their 2010 paper "Implementation Methods of Wall Functions in Cell-vertex Numerical Solvers".  The rest of the setup is identical to the ablNeutralEdge regression test._ 

and then just delete the rest of the paragraph

**Review by psakievich — COMMENTED** [review]
Looks pretty good to me.  I agree with @sayerhs's suggested changes.

**Review by psakievich — CHANGES_REQUESTED** [review]
Add a gold file for the regression test

File: reg_tests/test_files/ablNeutralEdgeNoSlip/README (hunk 1 above)
Comment (neilmatula) [reply reg_tests/test_files/ablNeutralEdgeNoSlip/README:37]: README was updated

File: src/LowMachEquationSystem.C (hunk 2 above)
Comment (neilmatula) [reply src/LowMachEquationSystem.C]: Done.

File: src/edge_kernels/MomentumABLWallShearStressEdgeKernel.C (hunk 4 above)
Comment (neilmatula) [reply src/edge_kernels/MomentumABLWallShearStressEdgeKernel.C]: Done.

File: src/ngp_algorithms/MomentumABLWallFuncMaskUtil.C (hunk 5 above)
Comment (neilmatula) [reply src/ngp_algorithms/MomentumABLWallFuncMaskUtil.C]: Outdated comment removed.

File: src/LowMachEquationSystem.C (hunk 6 above)
Comment (neilmatula) [reply src/LowMachEquationSystem.C]: Change made.

**Review by psakievich — APPROVED** [review]
